pennylane/devices/capabilities.py

```diff
@@ -0,0 +1,370 @@
+
+    Args:
+        document (dict): The TOML document loaded from a file.
+        prefix (str): Optional prefix corresponding to the runtime interface.
+
+    """
+
+    section = _get_toml_section(document, "compilation", prefix)
+
+    if unknowns := set(section) - VALID_COMPILATION_OPTIONS:
+        raise InvalidCapabilitiesError(
+            f"The compilation section has unknown options: {list(unknowns)}"
+        )
+
+    if not section.get("overlapping_observables", True) and section.get(
+        "non_commuting_observables", False
+    ):
+        raise InvalidCapabilitiesError(
+            "When overlapping_observables is False, non_commuting_observables cannot be True."
+        )
+
+    return section
+
+
+def _get_options(document: dict) -> dict[str, str]:
+    """Get custom options"""
+    return document.get("options", {})
+
+
+def parse_toml_document(document: dict) -> DeviceCapabilities:
+    """Parses a TOML document into a DeviceCapabilities object.
+
+    This function will ignore sections that are specific to either runtime interface, such as
+    "qjit.operators.gates". To include these sections, use :func:`update_device_capabilities`
+    on the capabilities object returned from this function.
+
+    """
+
+    schema = int(document["schema"])
+    assert schema in ALL_SUPPORTED_SCHEMAS, f"Unsupported capabilities TOML schema {schema}"
+    operations = _get_operations(document)
+    observables = _get_observables(document)
+    measurement_processes = _get_measurement_processes(document)
+    compilation_options = _get_compilation_options(document)
+    return DeviceCapabilities(
+        operations=operations,
+        observables=observables,
+        measurement_processes=measurement_processes,
+        **compilation_options,
+        options=_get_options(document),
+    )
+
+
+def update_device_capabilities(
+    capabilities: DeviceCapabilities, document: dict, runtime_interface: str
+):
+    """Updates the device capabilities objects with additions specific to the runtime interface."""
+
+    if runtime_interface not in {"pennylane", "qjit"}:
+        raise ValueError(f"Invalid runtime interface: {runtime_interface}")
+
+    operations = _get_operations(document, runtime_interface)
+    capabilities.operations.update(operations)
+
+    observables = _get_observables(document, runtime_interface)
+    capabilities.observables.update(observables)
+
+    measurement_processes = _get_measurement_processes(document, runtime_interface)
+    capabilities.measurement_processes.update(measurement_processes)
+
+    compilation_options = _get_compilation_options(document, runtime_interface)
+    for option, value in compilation_options.items():
+        setattr(capabilities, option, value)
+
+    if runtime_interface == "qjit" and "qjit" in document and not capabilities.qjit_compatible:
+        raise InvalidCapabilitiesError(
+            "qjit-specific sections are found but the device is not qjit-compatible."
+        )
```

Comment:
Wouldn't we want to raise this even if there are no "qjit specific" sections in the TOML file? I would think that's the most common thing for a device that isn't qjit compatible - that there are no qjit-specific sections (because no one has looked into how the device would interface with qjit), and they don't set the flag regarding qjit, so `qjit_compatible` is `False`.

Reply:
As discussed offline, if a device is not QJIT-compatible, the TOML file is still valid, so we choose not to raise an error here even if the TOML file is being loaded in a QJIT context. We leave it to Catalyst to raise the error if `qjit` is requested on a device that isn't QJIT-compatible.